packages/utils/src/lib/scoring.ts

```diff
@@ -77,27 +91,39 @@ export function calculateScore<T extends { weight: number }>(
   refs: T[],
   scoreFn: (ref: T) => number,
 ): number {
-  const numerator = refs.reduce(
-    (sum, ref) => sum + scoreFn(ref) * ref.weight,
-    0,
+  const { numerator, denominator } = refs.reduce(
+    (acc, ref) => {
+      acc.numerator += scoreFn(ref) * ref.weight;
+      acc.denominator += ref.weight;
+      return acc;
+    },
+    { numerator: 0, denominator: 0 },
   );
-  const denominator = refs.reduce((sum, ref) => sum + ref.weight, 0);
   return numerator / denominator;
 }
 
 export function scoreReport(report: Report): ScoredReport {
+  const allScoredAudits: EnrichedAuditReport[] = [];
+  const allScoredGroups: EnrichedScoredAuditGroup[] = [];
+
   const scoredPlugins = report.plugins.map(plugin => {
     const { groups, audits } = plugin;
     const preparedAudits = audits.map(audit => ({
       ...audit,
       plugin: plugin.slug,
     }));
+    allScoredAudits.push(...preparedAudits);
+
     const preparedGroups =
-      groups?.map(group => ({
-        ...group,
-        score: calculateScore(group.refs, groupRefToScore(preparedAudits)),
-        plugin: plugin.slug,
-      })) || [];
+      groups?.map(group => {
+        const scoredGroup = {
+          ...group,
+          score: calculateScore(group.refs, groupRefToScore(preparedAudits)),
+          plugin: plugin.slug,
+        };
+        allScoredGroups.push(scoredGroup);
+        return scoredGroup;
+      }) || [];
```

Comment:
Similar issue as before, but slightly different scenarios. Similar to `reduce`, `map` is intended to be used in an FP-style. Specifically, it should only return new array items, side-effects in `map` are unexpected (in this case, mutating arrays from outer scope). 

(Immutable/declarative methods are `map`, `filter`, `reduce`, `slice`, etc. While imperative/mutable methods are e.g. `forEach`, `sort`, `reverse`, `splice`.)

Reply:
I totally agree that using the imperative approach inside the declarative is not good. Rewrote the function, it became a bit longer, but is now fully declarative, with no mutation.